causallearn/utils/cit.py

```diff
@@ -88,7 +88,10 @@ def fisherz(self, X, Y, condition_set):
         """
         var = list((X, Y) + condition_set)
         sub_corr_matrix = self.correlation_matrix[np.ix_(var, var)]
-        inv = np.linalg.inv(sub_corr_matrix)
+        try:
+            inv = np.linalg.inv(sub_corr_matrix)
+        except np.linalg.LinAlgError:
```

Comment:
Are you sure all the np.linalg.LinAlgError will be singularity problem?

Also please make your description more informative: e.g. "data correlation matrix is singular. Cannot run fisherz test. Please check your data"

Reply:
Yes, all the np.linalg.LinAlgError in the function numpy.linalg.inv is the error of the singular matrix. Please refer to https://github.com/numpy/numpy/blob/v1.23.0/numpy/linalg/linalg.py#L483-L553.

Thanks for your suggestion! I will update it.

Reply:
where did you get this information?

I checked the link you gave, and it says:

    Raises
    ------
    LinAlgError
        If `a` is not square or inversion fails.
        
Does "inversion fails" guarantee "singular matrix"? I am not sure lol --- not very familiar it.

Reply:
Yes, the function np.linalg.inv gets error in [here](https://github.com/numpy/numpy/blob/v1.23.0/numpy/linalg/linalg.py#L551). The error is [singular matrix](https://github.com/numpy/numpy/blob/v1.23.0/numpy/linalg/linalg.py#L89). We can guarantee that the input matrix ’a‘ is a square matrix.

Reply:
awesome, this is clear. Thanks for the clear instructions.

I am curious, without your change, what would be the output if we have data with singular correlation? I think it would directly raise LinAlgError() with "Singular Matrix" as error message?

Do you mind running an example and attach the screenshot to your test plan?

Reply:
### Test plan section

Test fisherz under code version [5b30aad](https://github.com/cmu-phil/causal-learn/tree/5b30aada75b55b22f311ee514960f9f7558d245e). The test code is as follows.

``` python
import unittest

import numpy as np
from causallearn.utils.cit import CIT


class TestCIT(unittest.TestCase):
    def test_fisherz_singularity_problem(self):
        X1 = X2 = np.random.normal(size=1000)
        X = np.array([X1, X2]).T
        cit = CIT(data=X, method='fisherz')
        cit.fisherz(0, 1, tuple())
```

![image](https://user-images.githubusercontent.com/83860323/179175828-98cf8e9d-54f4-4112-80ca-36e3ea8ca152.png)

Reply:
I tried changing the inverse to pseudo-inverse and tested fisherz tests as the code below, and the image below appeared. r in the picture is the partial correlation coefficient statistic of the variables X1 and X2 in both cases. The partial correlation coefficient of the two variables in the first case is negative, which I think is wrong. I looked at the partial correlation coefficient as described on Wikipedia, and the inverse method does require the correlation coefficient matrix to be invertible. 
```python
import unittest

import numpy as np
from causallearn.utils.cit import CIT

class TestCIT(unittest.TestCase):
    def test_fisherz_singularity_problem(self):
        np.random.seed(767)
        
        X1 = X2 = np.random.normal(size=1000)
        X = np.array([X1, X2]).T
        cit = CIT(data=X, method='fisherz')
        cit.fisherz(0, 1, tuple())
        
        X1 = X1 + 0.01 * np.random.normal(size=1000)
        X2 = X2 + 0.01 * np.random.normal(size=1000)
        X = np.array([X1, X2]).T
        cit = CIT(data=X, method='fisherz')
        cit.fisherz(0, 1, tuple())
```

![image](https://user-images.githubusercontent.com/83860323/179349140-8499eab8-f9cb-483b-b48e-b3965d38333c.png)